apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.cs

```diff
@@ -255,13 +255,117 @@ public static SpannerCommandTextBuilder FromCommandText(string commandText)
                         commandText = $"{commandName} {targetTable}";
                     }
                     break;
             }
-            return new SpannerCommandTextBuilder(commandText, commandType, targetTable, extraStatements: null);
+            return new SpannerCommandTextBuilder(commandText.Trim(), commandType, targetTable, extraStatements: null);
+        }
+
+        private static string RemoveLeadingCommentsAndHints(string commandText, out bool removedCommentOrHint)
+        {
+            try
+            {
+                return commandText.Substring(IndexOfCommandStart(commandText, out removedCommentOrHint));
+            }
+            catch(Exception ex) when (ex is ArgumentOutOfRangeException || ex is IndexOutOfRangeException)
+            {
+                throw new ArgumentException($"'{commandText}' is not a recognized Spanner command.", nameof(commandText), ex);
+            }
+        }
+
+        private static int IndexOfCommandStart(string command, out bool removedCommentOrHint)
+        {
+            int currentIndex = 0;
+            int loops = 0;
+            do
+            {
+                loops++;
+                SkipWhiteSpaces();
+            }
+            while (SkipLineComment() || SkipBlockComment() || SkipHint());
+
+            removedCommentOrHint = loops > 1;
+            return currentIndex;
+
+            void SkipWhiteSpaces()
+            {
+                while (char.IsWhiteSpace(command[currentIndex]))
+                {
+                    currentIndex++;
+                }
+            }
+
+            bool SkipLineComment()
+            {
+                if (command.IndexOf('#', currentIndex, 1) >= 0 ||
+                    command.IndexOf("--", currentIndex, 2) >= 0)
+                {
+                    // TODO: If this can be escaped, let's check that it's not
+                    // or continue searching.
+                    currentIndex = command.IndexOf('\n', currentIndex);
+                    if (currentIndex >= 0) // If we didn't find it, we don't fail here.
+                    {
+                        currentIndex++;
+                    }
+                    return true;
+                }
+                return false;
+            }
+
+            bool SkipBlockComment()
+            {
+                if (command.IndexOf("/*", currentIndex, 2) >= 0 )
+                {
+                    // TODO: If this can be escaped, let's check that it's not
+                    // or continue searching.
+                    currentIndex = command.IndexOf("*/", currentIndex + 2);
+                    if (currentIndex >= 0) // If we didn't find it, we don't fail here.
+                    {
+                        currentIndex += 2;
+                    }
+                    return true;
+                }
+
+                return false;
+            }
+            bool SkipHint()
+            {
+                if (command.IndexOf('@', currentIndex, 1) >= 0)
```

Comment:
This is just `if (command[currentIndex] == '@')` isn't it?

Reply:
Same as above, This is just: we didn't find an end of block comment, so the whole thing is a comment. currentIndex will be -1 and the next time we try to use it it will throw ArgumentOutOfRange or IndexOutOfRange which are caught by the caller to throw ArgumentException("unrecognized command").

I'm fine with throwing here something more explicit though.